dhis-2/dhis-services/dhis-service-validation/src/main/java/org/hisp/dhis/outlierdetection/processor/ZscoreSqlStatementProcessor.java

```diff
@@ -166,34 +196,151 @@ public String getSqlStatement(OutlierDetectionRequest request) {
         + ";";
   }
 
-  /**
-   * To avoid the sql injection and decrease the load of the database engine (query plan caching)
-   * the named params are in use.
-   *
-   * @param request the instance of {@link OutlierDetectionRequest}.
-   * @return named params for parametrized sql query
-   */
-  @Override
-  public SqlParameterSource getSqlParameterSource(OutlierDetectionRequest request) {
-    return new MapSqlParameterSource()
-        .addValue(THRESHOLD.getKey(), request.getThreshold())
-        .addValue(DATA_ELEMENT_IDS.getKey(), request.getDataElementIds())
-        .addValue(START_DATE.getKey(), request.getStartDate())
-        .addValue(END_DATE.getKey(), request.getEndDate())
-        .addValue(DATA_START_DATE.getKey(), request.getDataStartDate())
-        .addValue(DATA_END_DATE.getKey(), request.getDataEndDate())
-        .addValue(MAX_RESULTS.getKey(), request.getMaxResults());
-  }
+  private String getModifiedZScoreSqlStatement(OutlierDetectionRequest request) {
+    String ouPathClause = OutlierDetectionUtils.getOrgUnitPathClause(request.getOrgUnits(), "ou");
+    String dataStartDateClause = getDataStartDateClause(request.getDataStartDate(), "", "t2");
+    String dataEndDateClause =
+        getDataEndDateClause(
+            request.getDataEndDate(), StringUtils.isEmpty(dataStartDateClause) ? "" : "and", "t2");
+    String order =
+        request.getOrderBy() == Order.MEAN_ABS_DEV
+            ? "middle_value_abs_dev"
+            : request.getOrderBy().getKey();
 
-  private String getDataStartDateClause(Date dataStartDate) {
-    return dataStartDate != null
-        ? "and pe.startdate >= :" + DATA_START_DATE.getKey() + " "
-        : StringUtils.EMPTY;
-  }
+    String thresholdParam = THRESHOLD.getKey();
 
-  private String getDataEndDateClause(Date dataStartDate) {
-    return dataStartDate != null
-        ? "and pe.enddate <= :" + DATA_END_DATE.getKey() + " "
-        : StringUtils.EMPTY;
+    return " with t2 as (select dv.dataelementid as dataelementid,"
```

Comment:
I think it could be formatted differently. I will ping you soon to chat about it.